toolchain/check/cpp/location.cpp

```diff
@@ -9,17 +9,49 @@
 
 namespace Carbon::Check {
 
-static auto GetFile(Context& context, SemIR::CheckIRId ir_id)
-    -> const SemIR::File* {
-  if (ir_id == context.sem_ir().check_ir_id()) {
-    // Common case: the IR is the current file.
-    return &context.sem_ir();
-  }
+namespace {
+struct FileInfo {
+  const SemIR::File* sem_ir;
+  clang::SourceLocation start_loc;
+};
+}  // namespace
+
+// Map a CheckIRId into information about the corresponding file in both SemIR
+// and Clang's source manager.
+static auto GetFileInfo(Context& context, SemIR::CheckIRId ir_id) -> FileInfo {
+  const SemIR::File* sem_ir = &context.sem_ir();
+  unsigned file_index = 0;
 
   // If the file is imported, locate it in our imports map.
-  auto import_id = context.check_ir_map().Get(ir_id);
-  CARBON_CHECK(import_id.has_value());
-  return context.import_irs().Get(import_id).sem_ir;
+  if (ir_id != context.sem_ir().check_ir_id()) {
+    auto import_id = context.check_ir_map().Get(ir_id);
+    CARBON_CHECK(import_id.has_value());
+    file_index = import_id.index + 1;
+
+    sem_ir = context.import_irs().Get(import_id).sem_ir;
+    CARBON_CHECK(sem_ir, "Node location in nonexistent IR");
+  }
+
+  // If we've seen this file before, reuse the same FileID.
+  auto& file_start_locs = context.cpp_carbon_file_locations();
+  if (file_start_locs.size() <= file_index) {
+    file_start_locs.resize(file_index + 1);
+  }
+  if (file_start_locs[file_index].isValid()) {
+    return {.sem_ir = sem_ir, .start_loc = file_start_locs[file_index]};
+  }
```

Comment:
Mainly suggesting for the `else`, to avoid the immediate `isValid` check.

```suggestion
  if (file_start_locs.size() <= file_index) {
    // Never valid; prepare a slot for the caching below.
    file_start_locs.resize(file_index + 1);
  } else if (file_start_locs[file_index].isValid()) {
    return {.sem_ir = sem_ir, .start_loc = file_start_locs[file_index]};
  }
```